addons/mail/models/mail_thread.py

```diff
@@ -3239,29 +3239,42 @@ def _notify_thread(self, message, msg_vals=False, **kwargs):
             set(kwargs.keys()),
             restricting_names=self._get_notify_valid_parameters()
         )
-
-        recipients_data = self._notify_get_recipients(message, msg_vals=msg_vals, **kwargs)
-        if not recipients_data:
-            return recipients_data
-        # cache data fetched by manual query to avoid extra queries when reading user.partner_id
-        uid2pid = {r['uid']: r['id'] for r in recipients_data if r['uid']}
-        users = self.env['res.users'].browse(uid2pid)
-        users._fields['partner_id']._insert_cache(users, uid2pid.values())
         # if scheduled for later: add in queue instead of generating notifications
         scheduled_date = self._is_notification_scheduled(kwargs.pop('scheduled_date', None))
         if scheduled_date:
             # send the message notifications at the scheduled date
-            self.env['mail.message.schedule'].sudo().create({
+            schedule = self.env['mail.message.schedule'].sudo().create({
                 'scheduled_datetime': scheduled_date,
                 'mail_message_id': message.id,
                 'notification_parameters': json.dumps(kwargs),
             })
+
+        # cache message data to avoid extra queries when accessing already-given values
+        cache_msg = self.env['mail.message'].browse(message.id)
+        for fname, fvalue in (msg_vals or {}).items():
+            cache_msg._fields[fname]._insert_cache(cache_msg, [fvalue])
```

Comment:
The only dangerous part is to store raw values rather than field post-processed values (for `body` in particular, user input must be sanitized).